### DISCUSSION TITLE
Designing a stateful `Form` widget

### DISCUSSION THREAD

**Discussion by jamesmcl113** [ratatui/ratatui#551] [Discussion]
Hi, I'm currently working on a small application with ratatui and I've found the need for some kind of Form-like widget for entering data into multiple fields. I've created a `StatefulWidget` impl for my Form based off the `Table` widget implementation, but with a slightly different interface to allow for different state when the user is selecting a field vs entering data into a field. I implemented this by using a `.select()` method on my custom form state which looks like this:
```rust
pub fn select(&mut self, s: Option<(usize, bool)>) {}
```
where the bool indicates whether the field is 'active' or not i.e. if text should be appended to the field or if it is just 'hovered'. This allows for more conditional rendering, as the form can have styles for each field depending on whether none are selected, a field is selected but not actively receiving charcters or if a field is active. This is the solution I came up with after some toying around and it seems to work fine, aswell as seeming like a natural extension of `TableState`'s interface for selection.

However, I also need some kind of interface that allows the user to submit the form so that we can get the values of the fields and do some style changes if e.g. a field is missing. After a lot more toying around I ended up adding a `.submit()` method which just toggles a flag, so that my rendering code can check if the form is submitted and some fields are empty => apply some styling to said fields. This works ok for my application, but I'm worried that I've coupled the state too tightly with the rendering code.

I'd eventually like to publish this widget for others to use as I think it could come in handy for a lot of different applications, so I'm a little concerned that the interface might be unwieldly and it won't be flexible enough for other users who might want to customize the behaviour of the form.

I'd be interested to know how people that have created similar types of widgets have handled this.

**Comment by joshka** [comment]
That's a cool idea. You might not need to take the stateful widget approach with this though. That's only really necessary to use if the rendering process has to have output that is due to the rendering (e.g. the table can adjust the selection to ensure that it's on the screen). Something that can also work just as well is to store a FormState or FormContext value in a field of the widget as a reference with a lifetime. See the tabs in demo2 for an example. Keeping the state separate lets you abstract and test the ideas of your form separate from the presentation of those ideas (e.g "when the user ... these fields have errors" vs "given this application state, draw these text boxes with a red background").

Regarding that bool. You might consider making a small enum to better communicate the meaning of what you wrote about active vs hover.

Just some thoughts, there's no right way to write any of this.

**Reply by jamesmcl113** [reply]
You're 100% right, that example looks like a much better way to implement what I want.
And for the bool, I think I could just have a `.status()` function that returns an enum which wraps the fields to make it clear if they're full, empty, some are empty etc.
Thanks for the response!

**Comment by jamesmcl113** [comment]
For anyone interested, I published the form widget on crates.io and the code is [here](https://github.com/jamesmcl113/tui_form_widget). Still missing some major things atm but definitely useable. I ended up embeding all the state within a `Form` struct that can return a widget (I took that from [tui-textarea](https://github.com/rhysd/tui-textarea) and it seems to work well) for rendering.

**Reply by joshka** [reply]
Perhaps add a screenshot of the example?
Take a look at how we use VHS in the tape files in Ratatui if you want some inspiration for how to do this easily.
I'd avoid keeping screenshots in the repo itself to avoid bloating your repo (instead upload the file in a comment and grab the link, or publish to the VHS site using `vhs blah.tape --publish`
